File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
      int len = 0;
      ListNode* temp = head;
      while(temp!=NULL){
        len++;
        temp = temp->next;
      }
      if(n==len){//this is for handling the case jha pe hme jis node ko htana ho whi agr first node hua toh agr temp hi null hoga to temp ka next to error show krega isliye hme isse handle krna pdega 
        head = head->next;
        return head;
      } 
      int m = len-n+1;
      int idx = m-1;
      temp = head;
      for(int i=1;i<=idx-1;i++){
        temp = temp->next;
      } 
      temp->next = temp->next->next;
      return head;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode();
        dummy.next = head; 
        ListNode slow = dummy;
        ListNode fast = dummy;
        // fast ko k steps aage le jao
        for(int i=0;i<=n;i++){
            fast = fast.next;
        }
        // ab dono ko saath me chalao
        while(fast!=null){
            slow = slow.next;
            fast = fast.next;
        }
        //remove nth node from end
        slow.next = slow.next.next;
        return dummy.next;
    }
}